custom/ewsghana/handlers/alerts.py

```diff
@@ -5,10 +6,173 @@
+                  or not mylist[i].isalnum() and mylist[i + 1].isalnum()):
+                yield token
+                token = ''
+            i += 1
+
+    def format(self, string):
+        """
+        Old parse method, used in Ghana for more 'interesting' parsing.
+        Moved from: https://github.com/dimagi/rapidsms-logistics/blob/7a1433abbda4ec27dc8f4c5da14c0f5689abd202/logistics/models.py#L1430
+        """
+        if not string:
+            return
+        match = re.search("[0-9]", string)
+        if not match:
+            return string
+        string = self._clean_string(string)
+        an_iter = self._getTokens(string)
+        commodity = None
+        valid = False
+
+        result = ''
+
+        while True:
+            try:
+                while commodity is None or not commodity.isalpha():
+                    commodity = an_iter.next().lower()
+                count = an_iter.next()
+                while not count.isdigit():
+                    count = an_iter.next()
+                result += '{} {}'.format(commodity, count)
+                valid = True
+                token_a = an_iter.next()
+                if not token_a.isalnum():
+                    token_b = an_iter.next()
+                    while not token_b.isalnum():
+                        token_b = an_iter.next()
+                    if token_b.isdigit():
+                        # if digit, then the user is reporting receipts
+                        result += '.{} '.format(token_b)
+                        commodity = None
+                        valid = True
+                    else:
+                        # if alpha, user is reporting soh, so loop
+                        commodity = token_b
+                        valid = True
+                else:
+                    commodity = token_a
+                    valid = True
+            except ValueError:
+                commodity = None
+                continue
+            except StopIteration:
+                break
+        if not valid:
+            return string
+        return result.strip()
+
+
+class EWSStockAndReceiptParser(StockAndReceiptParser):
+
+    def __init__(self, domain, v):
+        super(EWSStockAndReceiptParser, self).__init__(domain, v)
+        self.bad_codes = set()
+
+    def product_from_code(self, prod_code):
+        try:
+            return super(EWSStockAndReceiptParser, self).product_from_code(prod_code)
+        except SMSError:
+            return None
+
+    def single_action_transactions(self, action, args):
+        products = []
+        for idx, arg in enumerate(args):
+            if self.looks_like_prod_code(arg):
+                product = self.product_from_code(arg)
+                if product:
+                    products.append(product)
+                else:
+                    if idx == 0:
+                        raise ProductCodeException(INVALID_PRODUCT_CODE % arg)
```

Comment:
In EWS when first product is invalid exception is raised and other products aren't processed.